Sources/ChessKitEngineCore/EngineMessenger/EngineMessenger.mm

```diff
@@ -42,42 +43,64 @@ - (void)dealloc {
 }
 
 - (void)start {
-    _pipe = [NSPipe pipe];
-    _pipeReadHandle = [_pipe fileHandleForReading];
-    
-    dup2([[_pipe fileHandleForWriting] fileDescriptor], fileno(stdout));
-    
+
+
+    // set up read pipe
+    _readPipe = [NSPipe pipe];
+    _pipeReadHandle = [_readPipe fileHandleForReading];
+
+    dup2([[_readPipe fileHandleForWriting] fileDescriptor], fileno(stdout));
+
     [[NSNotificationCenter defaultCenter]
      addObserver:self
      selector:@selector(readStdout:)
      name:NSFileHandleReadCompletionNotification
      object:_pipeReadHandle
     ];
-    
+
     [_pipeReadHandle readInBackgroundAndNotify];
+
+    // set up write pipe
+    _writePipe = [NSPipe pipe];
+    _pipeWriteHandle = [_writePipe fileHandleForWriting];
+    dup2([[_writePipe fileHandleForReading] fileDescriptor], fileno(stdin));
+
+    // create command dispatch queue and start engine
+    _queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+
+    dispatch_async(_queue, ^{
+        _engine->initialize();
+    });
 }
 
 - (void)stop {
     [_pipeReadHandle closeFile];
-    
-    _pipe = NULL;
+    [_pipeWriteHandle closeFile];
+
+    _readPipe = NULL;
     _pipeReadHandle = NULL;
-    
+
+    _writePipe = NULL;
+    _pipeWriteHandle = NULL;
+
     [[NSNotificationCenter defaultCenter] removeObserver:self];
 }
 
 - (void)sendCommand: (NSString*) command {
-    _engine->send_command(std::string([command UTF8String]));
```

Comment:
You probably want to remove the send_command function from the Engine protocol altogether. But I see you mentioned already supporting stdin command sending for lc0 in the description so this is definitely something that needs to be done in order to make this mergeable.

Reply:
Yes once Lc0 is updated to the new method I will clean up and remove the old command execution